### ISSUE TITLE
League of Legends window breakage.

### ISSUE THREAD

**Issue by FuzzyQuills** [labwc/labwc#874]
When League of Legends launches the main game  after the client finalises champ select, the game window has several issues:
- Any toplevel windows owned by panels (aka. waybar) still show on top and are clickable.
- All other windows are always below the game window. (Except wofi, I'm guessing this is because wofi likely uses the overlay?)
- The game window completely lacks decorations when using windowed mode, making it impossible to exit the game in some cases.
- The window itself in all cases is completely missing in the window switcher, making it possible to interact with a window behind the game window using a keyboard.

This also has the potential to break the game's keyboard input as in one test I managed to jam the Ctrl key which meant I kept pinging with left click.

Current workarounds I've found to this issue:
- Run the game windowed, but only after unticking "let the window manager decorate the windows" in winecfg. This fixes the window switcher and is far less buggy in general.
- Use Gamescope. I haven't tested this.

**Comment by johanmalm** [comment]
If you rum with debug log (-d) do you get anything relating to this?? https://github.com/labwc/labwc/blob/d4288415c942bb3f8577f5fcddcaf49f199838f9/src/xwayland-unmanaged.c#L134

Also, we’ve not yet implemented: set_role, set_window_type and set_hints
https://github.com/labwc/labwc/blob/d4288415c942bb3f8577f5fcddcaf49f199838f9/include/xwayland.h#L38

Does it work with sway? If so, would you be able to do some printfs in the callbacks associated with those listeners to try to build a picture of what we need to do?

From what you’ve described it seems like we need to convert some unmanaged surface to a toplevel window.

**Comment by Consolatis** [comment]
> The window itself in all cases is completely missing in the window switcher, making it impossible to alt-tab back without clicking in the window.

This indeed sounds like an unmanaged window. I can't think of another reason why a window wouldn't show up in the window switcher. Either wayland native and layer-shell or xwayland and unmanaged.

**Comment by FuzzyQuills** [comment]
> If you rum with debug log (-d) do you get anything relating to this??
> 
> https://github.com/labwc/labwc/blob/d4288415c942bb3f8577f5fcddcaf49f199838f9/src/xwayland-unmanaged.c#L134
> 
> Also, we’ve not yet implemented: set_role, set_window_type and set_hints
> 
> https://github.com/labwc/labwc/blob/d4288415c942bb3f8577f5fcddcaf49f199838f9/include/xwayland.h#L38
> 
> Does it work with sway? If so, would you be able to do some printfs in the callbacks associated with those listeners to try to build a picture of what we need to do?
> 
> From what you’ve described it seems like we need to convert some unmanaged surface to a toplevel window.

I can confirm it works with Sway, tracing labwc now.

**Comment by FuzzyQuills** [comment]
Yep can confirm it's unmanaged window shenanigans; it reports override_redirect not handled.

It gets even weirder though; *if the game starts windowed,* it becomes a managed window and has the proper decorations and behaves as expected. Trying to switch to fullscreen has varying levels of breakage past that point. (I even managed to turn the League of Legends window into a 0x0px window trying to switch at some point)

**Comment by Consolatis** [comment]
Should be fixed by
- #883

Feel free to comment if its not the case.

**Comment by Consolatis** [comment]
Right, playing the game in \*any\* nested compositor (including labwc itself) should "fix" that. Reason being the nested compositor will not try to steal the focus, even if something within is asking for it. We can't really do anything about that on the labwc xwayland side without breaking potentially valid use-cases to request keyboard focus. At least not without turning labwc into a giant collection of app-workarounds.
